rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -167,27 +96,69 @@ public boolean decreaseObject(Poolable<T> obj)
     checkState(obj.getHost().equals(this.host),
             "Call to free object of wrong partition, current partition=%s requested partition = %s",
             this.host, obj.getHost());
-    objectRemoved();
     log.debug("Decreasing pool size object: " + obj);
     objectFactory.destroy(obj.getObject());
+    aliveObjectCount.decrementAndGet();
     obj.destroy();
     return true;
   }
 
-  private synchronized void objectRemoved()
+  public Poolable<T> getObject()
+  {
+    Poolable<T> object;
+    try {
+      takeSemaphore.tryAcquire(config.getMaxWaitMilliseconds(), TimeUnit.MILLISECONDS);
```

Comment:
Return value denotes timeout. If we timeout then we need to return instead of going ahead without permit, that can cause creating more than max connections for some time. At the time of returning we would destroy extra ones and then double decrement semaphore.
Please add UTs around this to capture different cases.

Reply:
Added logic to check the return type, if it is false (lock not acquired), return null. Also, added UT around the semaphore logic. Please have a look.